### ISSUE TITLE
Render blocking scripts

### ISSUE THREAD

**Issue by drlukacs** [joomlaworks/allvideos#37]
The plugin loads a number of "render blocking" scripts:

- plugins/content/jw_allvideos/jw_allvideos/includes/js/behaviour.js?v=4.8.0
- plugins/content/jw_allvideos/jw_allvideos/includes/js/wmvplayer/silverlight.js?v=4.8.0
- plugins/content/jw_allvideos/jw_allvideos/includes/js/wmvplayer/wmvplayer.js?v=4.8.0
- plugins/content/jw_allvideos/jw_allvideos/includes/js/quicktimeplayer/ac_quicktime.js?v=4.8.0
- plugins/content/jw_allvideos/jw_allvideos/includes/js/jwplayer/jwplayer.js?v=4.8.0

For some reason, these are not picked up by JCH, but more importantly, they are not marked as "deferred." This causes the page's  rating on Google Pagespeed to be lowered.

The cause of the Pagespeed issue is twofold:

1. The scripts are not marked as "deferred" (even though they appear to be working that way too).
2. The scripts are inserted in a way that makes JCH include them as separate, non-deferred scripts.

There is also a second, related problem: the scripts are not minified.

I would propose the following fix for the developers:
(a) minify the scripts using a standard tool; and
(b) add an option to have the scripts marked as "deferred."

**Comment by fevangelou** [comment]
There is an option to compress all these into a single file. Look in the plugin parameters.

**Comment by drlukacs** [comment]
@fevangelou Your response is incorrect and does not address the issues raised.

First, even when you have compression turned on, there will be several files generated:
- plugins/content/jw_allvideos/jw_allvideos/includes/js/jwp.js.php?v=4.8.0
- plugins/content/jw_allvideos/jw_allvideos/includes/js/jwplayer/jwplayer.js

Second these scripts are **"render blocking"**: see the report on any page using them in Google Page Speed tool. These scripts are NOT "minified" either (do you know understand the difference between "compressed" and "minified"?), and most importantly, they are NOT marked as "deferred" scripts.

I would propose the following fix for the developers:
(a) minify the scripts using a standard tool; and
(b) add an option to have the scripts marked as "deferred."

**Comment by drlukacs** [comment]
Hi Fotis,

Please re-open this issue. It is not resolved at all.

Gabor

**Comment by fevangelou** [comment]
There are JS snippets to initialize player blocks within the document. These get triggered only if the required JS code is already loaded. If the 2 scripts above are switched to "defer", this will cause issues (and therefore break) how player blocks are loaded.

**Comment by fevangelou** [comment]
Try it yourself. Make a demo of various video providers and file formats, then modify the $document->addScript() instances where the JS files are loaded and add the "defer" attribute.

More on that here: https://joomla.stackexchange.com/questions/4035/add-javascript-with-doc-addscript-with-async-true

**Comment by drlukacs** [comment]
I recall trying it with YouTube, and it does not seem to break anything there. To be clear, I made it "deferred" and not "async."

As a general rule, scripts should be running only after the page has been loaded -- so scripts should, to the extent possible, be "deferred" -- that is what Google encourages too.

**Comment by fevangelou** [comment]
YouTube is just an <iframe> tag. No JS involved in "triggering" the player block on Joomla's side. That's why it has no issues there.

Google has general recommendations. But those don't always apply. Especially in the context of a CMS which has certain rules for making enhancements to page output.

**Comment by drlukacs** [comment]
These recommendations do affect the page's score on Google Speed (and
thus, likely elsewhere).

**Comment by fevangelou** [comment]
Google PageSpeed is a test on client-side performance. And it's also a set of recommendations, not a rule.

Overall performance has at least 2 more factors, server-side performance and network congestion, which often play a more crucial role.

A few KBs added from AllVideos (which are BTW cached on subsequent page loads) will not kill your site's performance.
